libsolidity/ast/AST.cpp

```diff
@@ -132,6 +132,12 @@ FunctionDefinition const* ContractDefinition::constructor() const
 	return nullptr;
 }
 
+bool ContractDefinition::constructorIsPublic() const
+{
+	FunctionDefinition const* f = constructor();
+	return !f || f->isPublic();
```

Comment:
The contract can be constructed if it either does not have a constructor at all (i.e. the default one will be provided) or the constructor that is there is public.